internal/dag/httpproxy_processor.go

```diff
@@ -112,6 +112,10 @@ type HTTPProxyProcessor struct {
 	// configurable and off by default in order to support the feature
 	// without requiring all existing test cases to change.
 	SetSourceMetadataOnRoutes bool
+
+	// Allows cluster operators to prevent the processor from sorting the routes
+	// before sending them to Envoy.
+	ShouldSortRoutes bool
```

Comment:
I would like to try to keep this logic out of the various processors as much as possible and try to only make changes at the dag or xdscache level

e.g. we could have this sort of config on the `dag.Builder` that would get passed down to `dag.DAG.EnsureVirtualHost` and eventually `dag.VirtualHost.AddRoute` to abstract this away from the processors (which tbh should maybe be in their own package at this point)

theoretically the processors should already be adding routes in list-order, and I thought we already did process includes in depth-first order, based on this recursion: https://github.com/projectcontour/contour/blob/6917d85ce41f3050107994c06e6bb5b7039c99ed/internal/dag/httpproxy_processor.go#L726

Reply:
based on the different dag builder/DAG options, we could maybe keep both a list and map structures for routes on a vhost and only use one of the two based on the passed in config

just off the top of my head thoughts, not fully formed yet

Reply:
I am open to the idea, still need to marinate the ramifications on test coverage for this. That being said I am not sure 
how this would work because:

`EnsureVirtualHost` which effectively constructs the vhost is not aware of the caller. And we have 3 cases

* The caller is an HTTPProxy with controller in `omitRoute` mode: in which case it should set it to `false`
* The caller is an Ingress with controller in `omitRoute` mode: in which case it should set it to `true`
* The caller is an Ingress and an HTTPProxy with controller in `omitRoute` mode: in which case it should set it to `false`

So in that case it seems to me that it is the responsibility of the processor to communicate its preference. Basically the processor is `aware` of the conformance requirements and declares if it needs them for the `vhost` that it manages

Reply:
> theoretically the processors should already be adding routes in list-order, and I thought we already did process includes in depth-first order, based on this recursion:

😅 actually they werent because `expandPrefixMatches` was using a match internally and that re-ordered all routes